packages/firestore/src/model/mutation.ts

```diff
@@ -576,23 +661,83 @@ export class TransformMutation extends Mutation {
     const transformResults = [] as FieldValue[];
     for (const fieldTransform of this.fieldTransforms) {
       const transform = fieldTransform.transform;
-      if (transform instanceof ServerTimestampTransform) {
-        let previousValue: FieldValue | null = null;
 
-        if (baseDoc instanceof Document) {
-          previousValue = baseDoc.field(fieldTransform.field) || null;
-        }
+      let previousValue: FieldValue = null;
+      if (baseDoc instanceof Document) {
+        previousValue = baseDoc.field(fieldTransform.field) || null;
+      }
 
-        transformResults.push(
-          new ServerTimestampValue(localWriteTime, previousValue)
+      let transformResult: FieldValue = null;
+      if (transform instanceof ServerTimestampTransform) {
+        transformResult = new ServerTimestampValue(
+          localWriteTime,
+          previousValue
+        );
+      } else if (transform instanceof ArrayUnionTransformOperation) {
+        transformResult = this.arrayUnionTransformResult(
+          transform,
+          previousValue
+        );
+      } else if (transform instanceof ArrayRemoveTransformOperation) {
+        transformResult = this.arrayRemoveTransformResult(
+          transform,
+          previousValue
         );
       } else {
         return fail('Encountered unknown transform: ' + transform);
       }
+
+      transformResults.push(transformResult);
     }
     return transformResults;
   }
 
+  /**
+   * Transforms the provided `previousValue` via the provided `arrayUnion`. Used
+   * for both local application and after server acknowledgement.
+   */
+  private arrayUnionTransformResult(
+    arrayUnion: ArrayUnionTransformOperation,
+    previousValue: FieldValue | null
+  ): FieldValue {
+    const result = this.coercedFieldValuesArray(previousValue);
```

Comment:
Would it make sense to have this live within the Transform class itself? If TransformOperation had an abstract method, the code in `localTransformResults` and `serverTransformResults ` would not need to know about each classes semantics.

Reply:
I like that idea. I reworked this (including moving the TransformOperation classes to their own file to keep mutation.ts sane and to match iOS). I definitely like it better (and it reduced ~30 lines of code) though it's a _little_ wonky since 1) ServerTimestampTransform.transform() requires localWriteTime (which we only have during local-application) and 2) .transform() is called on the array transforms during "remote" application (after a write ack) when we don't have localWriteTime available, and so it has to be an optional param.